### PR TITLE
#164829377 Adds functionality to share an article using email and social media

### PR DESCRIPTION
**Pull request by Raywire**
**What does this PR do?**
Enables a user to share an article using email and social media platforms.

**Description of the tasks to be completed**

1. Add tests for social share functionality.

2. Add functionality to enable a user to share an article via email.

3. Add functionality to enable a  user to be able to share an article using social media on Facebook, Twitter, LinkedIn, Reddit and Telegram

**How should this be manually tested?**

- Open your terminal window and clone the repo,  and navigate to the root directory.
- Fetch all the branches and checkout into this feature's branch using the commands below
```bash
git fetch --all
git checkout ft-social-share-164829377 
```
- Create a virtual environment and then activate the environment by as shown below
```python
virtualenv venv
source venv/bin/activate
```
- Install all the required dependencies using pip by running 
```python
pip install -r requirements.txt
```
- Run the Django server to serve up the app by typing 
```python
python manage.py runserver
```
- Using postman, send a GET request containing the slug for the article you want to share and the social media provider
```localhost/api/articles/<slug>/share/<provider>/```

**Share article on Twitter**

Screenshot 2019-04-15 at 22 46 52

**Share article on Facebook**

Screenshot 2019-04-15 at 22 47 48

**Share article via email**

Screenshot 2019-04-15 at 22 48 33

**Share on Reddit**

Screenshot 2019-04-15 at 22 49 31

**Share on Telegram**

Screenshot 2019-04-15 at 22 50 15

**Share on LinkedIn**

Screenshot 2019-04-15 at 22 50 52

**Relevant PT stories**
https://www.pivotaltracker.com/story/show/164829377

### CODE REVIEW

**Review by kirega — DISMISSED** [review]
Solid implementation, as an area of improvement however is instead of using the if statements you could implement a switch case using a dictionary. What do you think about that?

**Comment by Raywire** [comment]
> Solid implementation, as an area of improvement however is instead of using the if statements you could implement a switch case using a dictionary. What do you think about that?



> Solid implementation, as an area of improvement however is instead of using the if statements you could implement a switch case using a dictionary. What do you think about that?

Thanks for the feedback I'll consider it.

**Review by kirega — DISMISSED** [review]
The switch statement is so much easier to read and I believe it will also run quicker than the if-statements. 
Good job.

**Review by LarryKarani — DISMISSED** [review]
Good work on the implementation. I have tested it locally works fine

**Review by ipaullly — DISMISSED** [review]
good implementation of the feature and the descriptive PR enabled me to test the functionality locally without any problems

File: authors/apps/articles/views.py [hunk 1]
```diff
@@ -16,7 +16,8 @@
 from authors.apps.authentication.models import User
 from .utils import is_article_owner, has_reviewed, round_average
 from .filters import ArticleFilter
-
+from .utils import generate_share_url
+import cloudinary
```
Comment (kirega) [inline authors/apps/articles/views.py]: I have noticed this cloudinary import is not used in your social share code and it is also a duplicate import in the file. Please remove this.

Comment (Raywire) [reply authors/apps/articles/views.py]: > I have noticed this cloudinary import is not used in your social share code and it is also a duplicate import in the file. Please remove this.

Thanks for pointing it out, I have removed the duplicate import.

**Review by WNjihia — APPROVED** [review]

**Review by LarryKarani — APPROVED** [review]
Good job implementing the changes. As far as the functionality is concerned everything works fine given I had tested it before and the changes you made do not affect the functionality in any way.

**Pull request merged**